samples: Bluetooth: peripheral_hr and central_hr extended advertising support

cvinayak opened the pull request:
Update peripheral_hr sample to support using extended
advertising and hence be long range compatible on SoCs
that support Bluetooth Low Energy Coded PHY.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: peripheral_hr extended advertising support

Update peripheral_hr sample to support using extended
advertising and hence be long range compatible on SoCs
that support Bluetooth Low Energy Coded PHY.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: central_hr extended advertising support

Update central_hr sample to support using extended
advertising and hence be long range compatible on SoCs
that support Bluetooth Low Energy Coded PHY.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: Add LED blinking to peripheral_hr

Add LED blinking state to peripheral_hr indicating
advertising state, and LED On state to indicate connected
to a central device, respectively.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: bsim: Bluetooth: central_hr peripheral_hr coverage

Add bsim test coverage for central_hr and peripheral_hr
samples.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

aescolar commented on samples/bluetooth/central_hr/sample.yaml:
```diff
@@ -5,3 +5,15 @@ tests:
+    platform_allow:
+      - nrf52_bsim
+      - nrf52dk/nrf52832
+      - nrf52840dk/nrf52840
```
nrf5340bsim_nrf5340_cpuapp too then :)

cvinayak committed:
Bluetooth: Controller: Default to advertising coding in ACL connection

Change default to using advertising coding in ACL connection
when establishing connections on Coded PHY.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add missing check for Coded PHY support

Add missing check for Coded PHY support when creating
extended advertising set.

Added a fallback in Controller for upper layers requesting
Coded PHY while Controller is implemented without Coded PHY.
In this case, 1M PHY will be used in extended advertising
for testing purposes.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on samples/bluetooth/peripheral_hr/overlay-phy_coded.conf:
```diff
@@ -0,0 +1,5 @@
+CONFIG_BT_AUTO_PHY_UPDATE=n
+CONFIG_BT_EXT_ADV=y
+
+CONFIG_BT_CTLR_ADV_DATA_LEN_MAX=191
```
Seems like we set `CONFIG_BT_CTLR_ADV_DATA_LEN_MAX=191` every time we do `CONFIG_BT_EXT_ADV=y`. Should we modify the default value of `CONFIG_BT_CTLR_ADV_DATA_LEN_MAX`?

Thalley commented on samples/bluetooth/peripheral_hr/src/main.c:
```diff
@@ -32,18 +33,28 @@ static const struct bt_data ad[] = {
 		      BT_UUID_16_ENCODE(BT_UUID_DIS_VAL))
 };
 
+/* Use atomic variable, 2 bits for connection and disconnection state */
+static ATOMIC_DEFINE(state, 2U);
+
+#define STATE_CONNECTED    1U
+#define STATE_DISCONNECTED 2U
```
What is the advantage of this over e.g. a `volatile bool`?

cvinayak replied:
As callback and the main threads are at different priority, use of atomic flags is context safe compared to use of `volatile bool`

Thalley commented on samples/bluetooth/peripheral_hr/src/main.c:
```diff
@@ -136,14 +132,73 @@ int main(void)
-	 * of starting delayed work so we do it here
-	 */
+
+#if !defined(CONFIG_BT_EXT_ADV)
+	printk("Starting Legacy Advertising (connectable and scannable)\n");
+	err = bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
+	if (err) {
+		printk("Advertising failed to start (err %d)\n", err);
+		return 0;
+	}
+
+#else /* CONFIG_BT_EXT_ADV */
+	struct bt_le_adv_param adv_param = {
+		.id = BT_ID_DEFAULT,
+		.sid = 0U,
+		.secondary_max_skip = 0U,
+		.options = (BT_LE_ADV_OPT_EXT_ADV |
+			    BT_LE_ADV_OPT_CONNECTABLE |
+			    BT_LE_ADV_OPT_CODED |
+			    BT_LE_ADV_OPT_USE_NAME),
+		.interval_min = BT_GAP_ADV_FAST_INT_MIN_2,
+		.interval_max = BT_GAP_ADV_FAST_INT_MAX_2,
+		.peer = NULL,
+	};
```
To me it makes more sense to use the same parameters, only add the `BT_LE_ADV_OPT_EXT_ADV` for the extended parameters. 

That way we can also use the same restart functionality for both configs.

cvinayak replied:
Will keep as-is as this sample shows the use of two different API parameters, use of public define versus local parameter variable.

Thalley commented on samples/bluetooth/peripheral_hr/src/main.c:
```diff
@@ -152,6 +207,25 @@ int main(void)
+			/* Connected callback executed */
+
+		} else if (atomic_test_and_clear_bit(state,
+						     STATE_DISCONNECTED)) {
```
This looks weird. 

When we are connected, state is set to `STATE_CONNECTED`, but we then clearing it here? So after one iteration, we are still connected, but `STATE_CONNECTED` is not set.

Thalley commented on samples/bluetooth/peripheral_hr/src/main.c:
```diff
@@ -152,6 +207,25 @@ int main(void)
+		} else if (atomic_test_and_clear_bit(state,
+						     STATE_DISCONNECTED)) {
+#if defined(CONFIG_BT_EXT_ADV)
+			printk("Starting Extended Advertising "
+			       "(connectable and non-scannable)\n");
+			err = bt_le_ext_adv_start(adv,
+						  BT_LE_EXT_ADV_START_DEFAULT);
+			if (err) {
+				printk("Failed to start extended advertising "
+				       "set (err %d)\n", err);
+				return 0;
+			}
```
Please use the recommended line length of `100`

cvinayak replied:
done.

Thalley commented on samples/bluetooth/peripheral_hr/prj.conf:
```diff
@@ -8,3 +8,5 @@ CONFIG_BT_BAS=y
 CONFIG_BT_HRS=y
 CONFIG_BT_DEVICE_NAME="Zephyr Heartrate Sensor"
 CONFIG_BT_DEVICE_APPEARANCE=833
+
+CONFIG_GPIO=y
```
Does enabling this for a board that does not have LED, increase memory usage without actually adding any new functionality?

cvinayak replied:
Not able to test, as all my boards have LED. But agree that logically there could be more code added to boards without LED when GPIO module being included.

Removed explicit enable in the prj.conf, but depend on boards that already enable GPIO be default, like most nRF DK boards. In this case, LED  will be used from devicetree, when enabled by default.

Thalley commented on samples/bluetooth/peripheral_hr/src/main.c:
```diff
@@ -122,10 +117,50 @@ static void hrs_notify(void)
+#define BLINK_ONOFF K_MSEC(500)
+
+static struct k_work_delayable blink_work;
+static bool                  led_is_on;
```
`led_is_on` seems to be modify by multiple threads. Should be `volatile` or have some sort of thread safety added

cvinayak replied:
The work is stopped (or not started) before the variable is modifed and hence it is safe.

Thalley commented on samples/bluetooth/central_hr/src/main.c:
```diff
@@ -168,15 +185,23 @@ static void start_scan(void)
 	if (err) {
-		printk("Scanning failed to start (err %d)\n", err);
-		return;
+		printk("Scanning with Coded PHY support failed "
+		       "(err %d)\n", err);
```
Please use recommended line length of `100`

Thalley commented on samples/bluetooth/central_hr/src/main.c:
```diff
@@ -153,9 +167,12 @@ static void device_found(const bt_addr_le_t *addr, int8_t rssi, uint8_t type,
 	if (type == BT_GAP_ADV_TYPE_ADV_IND ||
-	    type == BT_GAP_ADV_TYPE_ADV_DIRECT_IND) {
+	    type == BT_GAP_ADV_TYPE_ADV_DIRECT_IND ||
+	    type == BT_GAP_ADV_TYPE_EXT_ADV) {
```
In the case that it is extended, we should filter additionally on `BT_GAP_ADV_PROP_CONNECTABLE`

Thalley replied:
The comment above doesn't match the code. 

This will still call `bt_data_parse` on non-connectable extended advertising, right?

cvinayak replied:
yes, but that should be ok as the connection request is only sent to peripheral with HR service listed in its services AD type

Thalley replied:
> yes, but that should be ok as the connection request is only sent to peripheral with HR service listed in its services AD type

A device can advertise HR without being connectable. I don't see any reason why we wouldn't add the connectable check so that we don't have a device that may attempt to connect to non-connectable advertising

Thalley commented on samples/bluetooth/central_hr/boards/nrf5340bsim_nrf5340_cpuapp.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_BT_BUF_ACL_RX_SIZE=255
```
Why do we need this change?

Thalley commented on samples/bluetooth/central_hr/Kconfig.sysbuild:
```diff
@@ -0,0 +1,15 @@
+# Copyright 2023-2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+source "share/sysbuild/Kconfig"
+
+config NET_CORE_BOARD
+	string
+	default "nrf5340dk/nrf5340/cpunet" if "$(BOARD)" = "nrf5340dk"
+	default "nrf5340_audio_dk/nrf5340/cpunet" if "$(BOARD)" = "nrf5340_audio_dk"
+	default "nrf5340bsim_nrf5340_cpunet" if "$(BOARD)" = "nrf5340bsim_nrf5340_cpuapp"
+
+config NET_CORE_IMAGE_HCI_IPC
+	bool "HCI IPC image on network core"
+	default y
+	depends on NET_CORE_BOARD != ""
```
This seems unrelated to testing. Suggest to move to the earlier commit that adds nRF5340 support for the samples.

Thalley commented on tests/bsim/bluetooth/samples/central_hr_peripheral_hr/boards/nrf5340bsim_nrf5340_cpuapp.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_BT_BUF_ACL_RX_SIZE=255
```
Why do we need this both here and in the sample?

cvinayak replied:
updated with a comment in the file

cvinayak commented on tests/bsim/bluetooth/samples/compile.sh:
```diff
@@ -0,0 +1,48 @@
+  app=samples/bluetooth/peripheral_hr \
+    sysbuild=1 \
+    conf_overlay=overlay-phy_coded.conf \
+    exe_name=bs_${BOARD}_${app}_prj_conf \
```
@aescolar Now the name length is too long in the script file, should I reduce the name `central_hr_peripheral_hr` or do you have any suggestion so that the executable name does not execeed the 100 char column limit in the compliance?

aescolar replied:
Oh well.. silly checks
If you want you can replace:
`Execute ./bs_${BOARD}_tests_bsim_bluetooth_samples_central_hr_peripheral_hr_prj_conf_overlay-phy_coded_conf \`
with
`Execute ./bs_${BOARD_TS}_$(guess_test_long_name)_prj_conf_overlay-phy_coded_conf \`
(Note that due to hwmv2 now we need to use BOARD_TS to ensure we don't have "/" in the name.

aescolar commented:
@cvinayak how are we doing in this PR? Having some coverage for coded phy in CI would be very nice

Thalley commented:
> @cvinayak how are we doing in this PR? Having some coverage for coded phy in CI would be very nice

I think these samples are pretty common to run for first-time users - It would be great to have them tested in BSIM in CI :)

cvinayak commented:
> > @cvinayak how are we doing in this PR? Having some coverage for coded phy in CI would be very nice
> 
> I think these samples are pretty common to run for first-time users - It would be great to have them tested in BSIM in CI :)

Using these samples (Extended Scanning) on target has reveal an assertion and its analysis has caused the delay, hope to rework on this PR as soon as possible.

cvinayak committed:
samples: Bluetooth: Support sysbuild for central_hr and peripheral_hr

Add support for using sysbuild for central_hr and
peripheral_hr samples so that hci_ipc sample is built
for nRF53x SoCs boards.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

aescolar commented on tests/bsim/compile.source:
```diff
@@ -19,6 +19,7 @@ function _compile(){
   local app_root="${app_root:-${ZEPHYR_BASE}}"
   local BOARD_ROOT="${BOARD_ROOT:-${ZEPHYR_BASE}}"
   local conf_file="${conf_file:-prj.conf}"
+  local extra_conf_file="${extra_conf_file:-""}"
```
Nit: It would be nice to separate this file change in its own commit

cvinayak replied:
done

aescolar reviewed (dismissed):
Thanks @cvinayak a few changes are needed

cvinayak commented on samples/bluetooth/central_hr/Kconfig.sysbuild:
```diff
@@ -0,0 +1,15 @@
+	string
+	default "nrf5340dk/nrf5340/cpunet" if "$(BOARD)" = "nrf5340dk"
+	default "nrf5340_audio_dk/nrf5340/cpunet" if "$(BOARD)" = "nrf5340_audio_dk"
+	default "nrf5340bsim_nrf5340_cpunet" if "$(BOARD)" = "nrf5340bsim_nrf5340_cpuapp"
```
Done

cvinayak commented on samples/bluetooth/peripheral_hr/Kconfig.sysbuild:
```diff
@@ -0,0 +1,15 @@
+	string
+	default "nrf5340dk/nrf5340/cpunet" if "$(BOARD)" = "nrf5340dk"
+	default "nrf5340_audio_dk/nrf5340/cpunet" if "$(BOARD)" = "nrf5340_audio_dk"
+	default "nrf5340bsim_nrf5340_cpunet" if "$(BOARD)" = "nrf5340bsim_nrf5340_cpuapp"
```
done

cvinayak commented on tests/bsim/bluetooth/samples/central_hr_peripheral_hr/Kconfig.sysbuild:
```diff
@@ -0,0 +1,11 @@
+	int
+	# Let's pass the test arguments to the application MCU test
+	# otherwise by default they would have gone to the net core.
+	default 0 if $(BOARD) = "nrf5340bsim_nrf5340_cpuapp"
```
done

cvinayak commented on tests/bsim/bluetooth/samples/central_hr_peripheral_hr/tests_scripts/central_hr_peripheral_hr_extended.sh:
```diff
@@ -0,0 +1,30 @@
+# Copyright 2023-2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+# Simple selfchecking test for the central_hr/peripheral_hr samples,
```
Removed

cvinayak commented on tests/bsim/bluetooth/samples/compile.sh:
```diff
@@ -0,0 +1,76 @@
+#set -x #uncomment this line for debugging
+set -ue
+
+: "${BSIM_COMPONENTS_PATH:?BSIM_COMPONENTS_PATH must be defined}"
```
Removed

Thalley commented on samples/bluetooth/peripheral_hr/src/main.c:
```diff
@@ -126,24 +117,126 @@ static void hrs_notify(void)
+#endif
+#endif /* CONFIG_GPIO */
+
 int main(void)
```
`main` is now 150 lines of code. Suggest to move some of the code blocks into functions, so you have functions like

```
configure_gpio
start_advertising
start_led
stop_led
```

to make the flow of the `main` easier to follow.

Thalley commented on tests/bsim/bluetooth/samples/central_hr_peripheral_hr/tests_scripts/central_hr_peripheral_hr.sh:
```diff
@@ -0,0 +1,29 @@
+
+source ${ZEPHYR_BASE}/tests/bsim/sh_common.source
+
+EXECUTE_TIMEOUT=200
```
Why such a long timeout?

cvinayak commented on tests/bsim/bluetooth/samples/compile.sh:
```diff
@@ -0,0 +1,76 @@
+
+BOARD_ROOT="${BOARD_ROOT:-${ZEPHYR_BASE}}"
+
+mkdir -p ${WORK_DIR}
```
Removed

cvinayak commented on tests/bsim/bluetooth/samples/compile.sh:
```diff
@@ -0,0 +1,76 @@
+
+source ${ZEPHYR_BASE}/tests/bsim/compile.source
+
+if [ "${BOARD_TS}" == "nrf5340bsim_nrf5340_cpuapp" ]; then
```
Removed duplicates

cvinayak committed:
tests: bsim: compile: Add support to supply EXTRA_CONF_FILE

Add support in bsim compile script to accept EXTRA_CONF_FILE
so that overlay conf files can be supplied at compile time.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on tests/bsim/bluetooth/samples/central_hr_peripheral_hr/src/sample_test.c:
```diff
@@ -0,0 +1,76 @@
+
+#define WAIT_TIME 120 /* Seconds */
+
+#define PASS_THRESHOLD 100 /* packets */
```
```suggestion
#define PASS_THRESHOLD 5 /* packets */
```

I don't see any reason to wait for 100 notifications. The behavior should be verifiable with a few notifications

Thalley commented on tests/bsim/bluetooth/samples/central_hr_peripheral_hr/src/sample_test.c:
```diff
@@ -0,0 +1,76 @@
+static const struct bst_test_instance test_sample[] = {
+	{
+		.test_id = "central_hr_peripheral_hr",
+		.test_descr = "Test based on a sample. "
```
```suggestion
		.test_descr = "Test based on the peripheral and central HR samples. "
```

cvinayak replied:
missed this one... pushed again.

Thalley reviewed (dismissed):
The code is OK, but I would heavily suggest to reduce the runtime of the tests
